From 7093aab8d1b2d1fd103c65ce93fa59e3da57c83e Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius-cristian.vlad@nxp.com>
Date: Thu, 7 Jun 2018 16:47:34 +0300
Subject: [PATCH 4028/5242] MGS-3969 [#imx-879] Fix reading HW counters values
 using old profiling method for GC7000

commit  037e57aeab0c4e5853504a5e8e62a5a437eb5282 from
https://source.codeaurora.org/external/imx/linux-imx.git

Adapted from 7a3d1096aa0cc77e258. This brings back the ability for
gputop to read HW counter values.

Signed-off-by: Marius Vlad <marius-cristian.vlad@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c     |   19 ++++++++++++++++++-
 .../mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c |   13 -------------
 2 files changed, 18 insertions(+), 14 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
index efb0159..2cdc21f 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
@@ -2486,6 +2486,7 @@
         }
         else
         {
+	    gctUINT32 i;
             if (Interface->u.Commit.count > 1 && Interface->engine == gcvENGINE_RENDER)
             {
                 gctUINT32 i;
@@ -2530,7 +2531,6 @@
 
             if (Interface->u.Commit.count > 1 && Interface->engine == gcvENGINE_RENDER)
             {
-                gctUINT32 i;
 
                 for (i = 1; i < Interface->u.Commit.count; i++)
                 {
@@ -2567,6 +2567,23 @@
                     }
                 }
             }
+
+	    for (i = 0; i < Interface->u.Commit.count; i++) {
+		    gceHARDWARE_TYPE type = Interface->hardwareType;
+		    gckKERNEL kernel = Device->map[type].kernels[i];
+
+		    if  ((kernel->hardware->options.gpuProfiler == gcvTRUE) &&
+		         (kernel->profileEnable == gcvTRUE)) {
+			    gcmkONERROR(gckCOMMAND_Stall(kernel->command, gcvTRUE));
+
+			    if (kernel->command->currContext) {
+				    gcmkONERROR(gckHARDWARE_UpdateContextProfile(
+							    kernel->hardware,
+							    kernel->command->currContext));
+			    }
+		    }
+	    }
+
         }
         gcmkONERROR(gckOS_ReleaseMutex(Kernel->os, Kernel->device->commitMutex));
         commitMutexAcquired = gcvFALSE;
diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
index 7a65118..a9cb300 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
@@ -2538,19 +2538,6 @@
     gcmkONERROR(gckCOMMAND_ExitCommit(Command, gcvFALSE));
     commitEntered = gcvFALSE;
 
-    if  ((Command->kernel->hardware->options.gpuProfiler == gcvTRUE) &&
-         (Command->kernel->profileEnable == gcvTRUE))
-    {
-        gcmkONERROR(gckCOMMAND_Stall(Command, gcvTRUE));
-
-        if (Command->currContext)
-        {
-            gcmkONERROR(gckHARDWARE_UpdateContextProfile(
-                        hardware,
-                        Command->currContext));
-        }
-    }
-
     if (status == gcvSTATUS_INTERRUPTED)
     {
         gcmkTRACE(
-- 
1.7.9.5

